### PR TITLE
feat(environment): support custom environment

### LINKED ISSUE #1484: [Feature]: Support for Custom Environment Tags
### Is your feature request related to a problem?

Now, ODC provides four environment tags: Development, Testing, Production, and Default. However, I also require additional environment tags, such as Performance Testing. It is hoped that ODC can offer the capability to customize environment tags in order to meet practical usage demands.

### Describe the solution you'd like

It is hoped that ODC can offer the capability to customize environment tags in order to meet practical usage demands.

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by MarkPotato777**
#### What type of PR is this?

type-feature
module-environment

#### What this PR does / why we need it:
This PR added a new feature that could let users be able to add a new environment based on an existing environment. (inherit the existing environment's sql-console and sql-check rules).
#### Which issue(s) this PR fixes:

Fixes #1484 

#### Additional documentation e.g., usage docs, etc.:

```docs

```

### CODE REVIEW

File: server/odc-migrate/src/main/resources/migrate/common/V_4_2_4_1__alter_collaboration_environment.sql [hunk 1]
```diff
@@ -0,0 +1 @@
+alter table `collaboration_environment` add column `is_enabled` tinyint(1) not null default '1' comment 'is the environemnt enabled';
```
Comment (smallsheeeep) [inline server/odc-migrate/src/main/resources/migrate/common/V_4_2_4_1__alter_collaboration_environment.sql:1]: Version number conflict, may use 4_2_4__7

File: server/odc-server/src/main/java/com/oceanbase/odc/server/web/controller/v2/EnvironmentController.java [hunk 2]
```diff
@@ -53,4 +59,28 @@ public SuccessResponse<Environment> getEnvironment(@PathVariable Long id) {
         return Responses.success(environmentService.detail(id));
     }
 
+    @ApiOperation(value = "createEnvironment", notes = "Create an environment")
+    @RequestMapping(value = "/environments", method = RequestMethod.POST)
+    public SuccessResponse<Environment> createEnvironment(@RequestBody @Valid CreateEnvironmentReq req) {
```
Comment (smallsheeeep) [inline server/odc-server/src/main/java/com/oceanbase/odc/server/web/controller/v2/EnvironmentController.java:64]: Should add `@Validated` for the controller class?

File: server/odc-service/src/main/java/com/oceanbase/odc/metadb/collaboration/EnvironmentRepository.java [hunk 3]
```diff
@@ -30,4 +31,7 @@ public interface EnvironmentRepository extends JpaRepository<EnvironmentEntity,
 
     Optional<EnvironmentEntity> findByIdAndOrganizationId(Long id, Long organizationId);
 
+    @Query(value = "update collaboration_environment set is_enabled=:enabled where id=:id", nativeQuery = true)
```
Comment (smallsheeeep) [inline server/odc-service/src/main/java/com/oceanbase/odc/metadb/collaboration/EnvironmentRepository.java:34]: Should add `@Param()` for native query?

File: server/odc-service/src/main/java/com/oceanbase/odc/service/collaboration/environment/EnvironmentService.java [hunk 4]
```diff
@@ -63,6 +83,32 @@ public class EnvironmentService {
...
+    @Autowired
+    private ConnectionService connectionService;
+
+    @Transactional(rollbackFor = Exception.class)
+    @PreAuthenticate(actions = "create", resourceType = "ODC_ENVIRONMENT", isForAll = true)
+    public Environment create(@NotNull CreateEnvironmentReq req) {
+        PreConditions.validNoDuplicated(ResourceType.ODC_ENVIRONMENT, "name", req.getName(),
+                () -> exists(req.getName()));
+
+        Ruleset saveedRuleset = rulesetService.create(buildRuleset(req.getName(), req.getDescription()));
```
Comment (smallsheeeep) [inline server/odc-service/src/main/java/com/oceanbase/odc/service/collaboration/environment/EnvironmentService.java:105]: `saveedRuleset` typo

File: server/odc-service/src/main/java/com/oceanbase/odc/service/collaboration/environment/EnvironmentService.java [hunk 5]
```diff
@@ -102,29 +148,89 @@ public Environment detailSkipPermissionCheck(@NonNull Long id) {
 
     @PreAuthenticate(actions = "update", resourceType = "ODC_ENVIRONMENT", indexOfIdParam = 0)
     @Transactional(rollbackFor = Exception.class)
-    public Environment update(@NonNull Long id, @NonNull Environment environment) {
-        EnvironmentEntity previous = environmentRepository.findByIdAndOrganizationId(id,
-                authenticationFacade.currentOrganizationId())
-                .orElseThrow(() -> new NotFoundException(ResourceType.ODC_ENVIRONMENT, "id", id));
+    public Environment update(@NotNull Long id, @NotNull UpdateEnvironmentReq req) {
```
Comment (smallsheeeep) [inline server/odc-service/src/main/java/com/oceanbase/odc/service/collaboration/environment/EnvironmentService.java:151]: `@Validated` should be add for this class.

File: server/odc-service/src/main/java/com/oceanbase/odc/service/collaboration/environment/EnvironmentService.java [hunk 6]
```diff
@@ -102,29 +148,89 @@ public Environment detailSkipPermissionCheck(@NonNull Long id) {
...
+        }
+        environmentRepository.deleteById(id);
+        rulesetService.delete(environment.getRulesetId());
+        return environment;
+    }
+
+
+    private void checkDataSourceReference(Long id) {
+        Set<ConnectionConfig> referencedConnections =
+                connectionService.listByOrganizationId(authenticationFacade.currentOrganizationId()).stream()
```
Comment (smallsheeeep) [inline server/odc-service/src/main/java/com/oceanbase/odc/service/collaboration/environment/EnvironmentService.java:191]: Why not list by environment id directly? May better performance.

File: server/odc-service/src/main/java/com/oceanbase/odc/service/collaboration/environment/model/CreateEnvironmentReq.java [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+ * @Date: 2024/1/11 10:17
+ * @Description: []
+ */
+@Data
+@Builder
+@NoArgsConstructor
+@AllArgsConstructor
+public class CreateEnvironmentReq {
+    @Name
+    @Size(min = 1, max = 5, message = "The length of the environment name must be between 1 and 5")
```
Comment (smallsheeeep) [inline server/odc-service/src/main/java/com/oceanbase/odc/service/collaboration/environment/model/CreateEnvironmentReq.java:39]: Why the max length is only5? The built-in "default" env is 7.

File: server/odc-service/src/main/java/com/oceanbase/odc/service/collaboration/environment/model/UpdateEnvironmentReq.java [hunk 8]
```diff
@@ -0,0 +1,35 @@
...
+ * @Date: 2024/1/12 14:40
+ * @Description: []
+ */
+@Data
+public class UpdateEnvironmentReq {
+    @Size(max = 256, message = "The length of the environment description must be between 0 and 256")
+    private String description;
+
+    @NotNull
+    private EnvironmentStyle style;
```
Comment (smallsheeeep) [inline server/odc-service/src/main/java/com/oceanbase/odc/service/collaboration/environment/model/UpdateEnvironmentReq.java:34]: The `EnvironmentStyle` may add more elements for more color style?

File: server/odc-service/src/main/java/com/oceanbase/odc/service/regulation/ruleset/RuleService.java [hunk 9]
```diff
@@ -189,6 +191,34 @@ public Rule update(@NonNull Long rulesetId, @NonNull Long ruleId, @NonNull Rule
...
+            entity.setLevel(rule.getLevel());
+            entity.setEnabled(rule.getEnabled());
+            entity.setPropertiesJson(JsonUtils.toJson(rule.getProperties()));
+            if (Objects.nonNull(rule.getAppliedDialectTypes())) {
+                entity.setAppliedDialectTypes(
+                        rule.getAppliedDialectTypes().stream().map(DialectType::name).collect(Collectors.toList()));
+            }
+            return entity;
+        }).collect(Collectors.toList());
+        ruleApplyingRepository.saveAll(entities);
```
Comment (smallsheeeep) [inline server/odc-service/src/main/java/com/oceanbase/odc/service/regulation/ruleset/RuleService.java:211]: Better to use batchCreate for better performance, refer to `OdcJpaRepository`.

File: server/odc-service/src/main/java/com/oceanbase/odc/service/regulation/ruleset/RulesetService.java [hunk 10]
```diff
@@ -57,7 +61,27 @@ public Ruleset detail(@NonNull Long id) {
         return ruleset;
     }
 
+    @Transactional(rollbackFor = Exception.class)
+    public Ruleset create(@NonNull Ruleset ruleset) {
```
Comment (smallsheeeep) [inline server/odc-service/src/main/java/com/oceanbase/odc/service/regulation/ruleset/RulesetService.java:65]: Lack of unit tests.

File: server/odc-migrate/src/main/resources/migrate/common/V_4_2_4_1__alter_collaboration_environment.sql (hunk 1 above)
Comment (MarkPotato777) [reply server/odc-migrate/src/main/resources/migrate/common/V_4_2_4_1__alter_collaboration_environment.sql]: done~

File: server/odc-service/src/main/java/com/oceanbase/odc/metadb/collaboration/EnvironmentRepository.java (hunk 3 above)
Comment (MarkPotato777) [reply server/odc-service/src/main/java/com/oceanbase/odc/metadb/collaboration/EnvironmentRepository.java:37]: Yes you are right. Updated.

File: server/odc-service/src/main/java/com/oceanbase/odc/service/collaboration/environment/EnvironmentService.java (hunk 4 above)
Comment (MarkPotato777) [reply server/odc-service/src/main/java/com/oceanbase/odc/service/collaboration/environment/EnvironmentService.java]: ❤️ thanks for your kind reminder

File: server/odc-service/src/main/java/com/oceanbase/odc/service/collaboration/environment/model/UpdateEnvironmentReq.java (hunk 8 above)
Comment (MarkPotato777) [reply server/odc-service/src/main/java/com/oceanbase/odc/service/collaboration/environment/model/UpdateEnvironmentReq.java:36]: Yes, added.

File: server/odc-service/src/main/java/com/oceanbase/odc/service/regulation/ruleset/RulesetService.java (hunk 10 above)
Comment (MarkPotato777) [reply server/odc-service/src/main/java/com/oceanbase/odc/service/regulation/ruleset/RulesetService.java:70]: It lacks unit tests as this was a WIP PR when you were reviewing. Now I added the unit cases, thanks for your reminder.

File: server/odc-service/src/main/java/com/oceanbase/odc/service/collaboration/environment/model/CreateEnvironmentReq.java (hunk 7 above)
Comment (MarkPotato777) [reply server/odc-service/src/main/java/com/oceanbase/odc/service/collaboration/environment/model/CreateEnvironmentReq.java]: I set it to 8 now. The max length is not supposed to be set to a huge number as the name would be shown in many places like SQL console. If the name length is too large, the UI will look like a kind of inharmonious. And if users DO need a longer environment name, we may adjust it in a later version. It's easy to make a short column be a long one, but not vice versa.

File: server/odc-server/src/main/java/com/oceanbase/odc/server/web/controller/v2/EnvironmentController.java (hunk 2 above)
Comment (MarkPotato777) [reply server/odc-server/src/main/java/com/oceanbase/odc/server/web/controller/v2/EnvironmentController.java:68]: Just adding `@Validated` on the class level will not help you validate all the methods. You still need to add `@Valid` or `@Validated` in the front of method parameters.

File: server/odc-service/src/main/java/com/oceanbase/odc/service/collaboration/environment/EnvironmentService.java (hunk 6 above)
Comment (MarkPotato777) [reply server/odc-service/src/main/java/com/oceanbase/odc/service/collaboration/environment/EnvironmentService.java]: You are right! Updated.

File: server/odc-service/src/main/java/com/oceanbase/odc/service/regulation/ruleset/RuleService.java (hunk 9 above)
Comment (MarkPotato777) [reply server/odc-service/src/main/java/com/oceanbase/odc/service/regulation/ruleset/RuleService.java:237]: Appreciate your reminder! I use `OdcJpaRepositiry#batchCreate` instead :)

File: server/odc-service/src/main/java/com/oceanbase/odc/metadb/collaboration/EnvironmentSpecs.java [hunk 11]
```diff
@@ -0,0 +1,45 @@
...
+public class EnvironmentSpecs {
+    public static Specification<EnvironmentEntity> organizationIdEquals(Long organizationId) {
+        return columnEquals("organizationId", organizationId);
+    }
+
+    public static Specification<EnvironmentEntity> enabledEquals(Boolean enabled) {
+        return columnEquals("enabled", enabled);
+    }
+
+    private static Specification<EnvironmentEntity> columnEquals(String column, Object value) {
```
Comment (smallsheeeep) [inline server/odc-service/src/main/java/com/oceanbase/odc/metadb/collaboration/EnvironmentSpecs.java:36]: You can use `SpecificationUtil` for less codes.

File: server/odc-service/src/main/java/com/oceanbase/odc/metadb/regulation/ruleset/RuleApplyingRepository.java [hunk 12]
```diff
@@ -43,5 +47,30 @@ Optional<RuleApplyingEntity> findByOrganizationIdAndRulesetIdAndRuleMetadataId(L
...
+    int deleteByOrganizationIdAndRulesetId(Long organizationId, Long rulesetId);
+
+
+    default List<RuleApplyingEntity> batchCreate(List<RuleApplyingEntity> entities) {
+        String sql = InsertSqlTemplateBuilder.from("regulation_rule_applying")
+                .field(RuleApplyingEntity_.ORGANIZATION_ID)
+                .field(RuleApplyingEntity_.RULESET_ID)
+                .field(RuleApplyingEntity_.RULE_METADATA_ID)
+                .field(RuleApplyingEntity_.APPLIED_DIALECT_TYPES)
+                .field(RuleApplyingEntity_.ENABLED)
```
Comment (smallsheeeep) [inline server/odc-service/src/main/java/com/oceanbase/odc/metadb/regulation/ruleset/RuleApplyingRepository.java:60]: The value of  `RuleApplyingEntity_.ENABLED` is "enabled", but the actual column name is "is_enabled". Have you ever test this case?

File: server/odc-service/src/main/java/com/oceanbase/odc/metadb/regulation/ruleset/RuleApplyingRepository.java [hunk 13]
```diff
@@ -43,5 +47,30 @@ Optional<RuleApplyingEntity> findByOrganizationIdAndRulesetIdAndRuleMetadataId(L
...
+                .field(RuleApplyingEntity_.LEVEL)
+                .field(RuleApplyingEntity_.PROPERTIES_JSON)
+                .build();
+        List<Function<RuleApplyingEntity, Object>> getter = valueGetterBuilder().add(
+                RuleApplyingEntity::getOrganizationId)
+                .add(RuleApplyingEntity::getRulesetId)
+                .add(RuleApplyingEntity::getRuleMetadataId)
+                .add(RuleApplyingEntity::getAppliedDialectTypes)
```
Comment (smallsheeeep) [inline server/odc-service/src/main/java/com/oceanbase/odc/metadb/regulation/ruleset/RuleApplyingRepository.java:68]: The `appliedDialectTypes` is a List, I am not sure whether the `batchCreate` method support this. You better to test that.

File: server/odc-service/src/main/java/com/oceanbase/odc/service/regulation/ruleset/RuleService.java [hunk 14]
```diff
@@ -156,6 +168,22 @@ public Rule detail(@NonNull Long rulesetId, @NonNull Long ruleId) {
 
     @PreAuthenticate(actions = "update", resourceType = "ODC_RULESET", indexOfIdParam = 0)
     public Rule update(@NonNull Long rulesetId, @NonNull Long ruleId, @NonNull Rule rule) {
```
Comment (smallsheeeep) [inline server/odc-service/src/main/java/com/oceanbase/odc/service/regulation/ruleset/RuleService.java:170]: `@Transactional` lost.

File: server/odc-service/src/main/java/com/oceanbase/odc/metadb/regulation/ruleset/RuleApplyingRepository.java (hunk 12 above)
Comment (MarkPotato777) [reply server/odc-service/src/main/java/com/oceanbase/odc/metadb/regulation/ruleset/RuleApplyingRepository.java:60]: You don't need to care about what the actual column name in tables, you only need to focus on the entity itself.

File: server/odc-service/src/main/java/com/oceanbase/odc/service/regulation/ruleset/RuleService.java (hunk 14 above)
Comment (MarkPotato777) [reply server/odc-service/src/main/java/com/oceanbase/odc/service/regulation/ruleset/RuleService.java:170]: okay, done

File: server/odc-service/src/main/java/com/oceanbase/odc/metadb/collaboration/EnvironmentSpecs.java (hunk 11 above)
Comment (MarkPotato777) [reply server/odc-service/src/main/java/com/oceanbase/odc/metadb/collaboration/EnvironmentSpecs.java:36]: Nice advice! Updated.

File: server/odc-service/src/main/java/com/oceanbase/odc/metadb/regulation/ruleset/RuleApplyingRepository.java (hunk 12 above)
Comment (MarkPotato777) [reply server/odc-service/src/main/java/com/oceanbase/odc/metadb/regulation/ruleset/RuleApplyingRepository.java:60]: Thanks for your careful review! Yes you are right, we should use `is_enabled` here. Plus, I added unit test cases for this method.

File: server/odc-service/src/main/java/com/oceanbase/odc/metadb/regulation/ruleset/RuleApplyingRepository.java (hunk 13 above)
Comment (MarkPotato777) [reply server/odc-service/src/main/java/com/oceanbase/odc/metadb/regulation/ruleset/RuleApplyingRepository.java:68]: Yes you are right, we need to convert the list to a string. Thanks for your kind reminder:)

File: server/odc-service/src/main/java/com/oceanbase/odc/metadb/regulation/ruleset/RuleApplyingRepository.java (hunk 12 above)
Comment (smallsheeeep) [reply server/odc-service/src/main/java/com/oceanbase/odc/metadb/regulation/ruleset/RuleApplyingRepository.java]: 😘

**Review by smallsheeeep — APPROVED** [review]
LGTM

**Review by smallsheeeep — DISMISSED** [review]

**Review by smallsheeeep — DISMISSED** [review]

File: server/odc-migrate/src/main/resources/migrate/common/V_4_2_4_3__alter_collaboration_environment.sql [hunk 15]
```diff
@@ -0,0 +1 @@
+alter table `collaboration_environment` add column `is_enabled` tinyint(1) not null default '1' comment 'is the environemnt enabled';
```
Comment (yhilmare) [inline server/odc-migrate/src/main/resources/migrate/common/V_4_2_4_3__alter_collaboration_environment.sql:1]: 4.2.4.3 is occupied

File: server/odc-migrate/src/main/resources/migrate/common/V_4_2_4_7__iam_permission.yaml [hunk 16]
```diff
@@ -0,0 +1,78 @@
+kind: resource
```
Comment (yhilmare) [inline server/odc-migrate/src/main/resources/migrate/common/V_4_2_4_7__iam_permission.yaml:1]: 4.2.4.7 is occupied

Comment (MarkPotato777) [reply server/odc-migrate/src/main/resources/migrate/common/V_4_2_4_7__iam_permission.yaml:1]: done

File: server/odc-migrate/src/main/resources/migrate/common/V_4_2_4_3__alter_collaboration_environment.sql (hunk 15 above)
Comment (MarkPotato777) [reply server/odc-migrate/src/main/resources/migrate/common/V_4_2_4_3__alter_collaboration_environment.sql]: done

**Review by yhilmare — APPROVED** [review]
lgtm

**Pull request merged**